call_test.go

```diff
@@ -213,6 +215,187 @@ func setUp(t *testing.T, port int, maxStreams uint32) (*server, *ClientConn) {
+		return invoker(secondCtx, method, req, reply, cc, opts...)
+	}
+
+	lastInt := func(ctx context.Context, method string, req, reply interface{}, cc *ClientConn, invoker UnaryInvoker, opts ...CallOption) error {
+		if ctx.Value(parentKey) == nil {
+			t.Fatalf("last interceptor should have %v in context", parentKey)
+		}
+		if ctx.Value(firstIntKey) == nil {
+			t.Fatalf("last interceptor should have %v in context", firstIntKey)
+		}
+		if ctx.Value(secondIntKey) == nil {
+			t.Fatalf("last interceptor should have %v in context", secondIntKey)
+		}
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+
+	server, cc := setUpWithOptions(t, 0, math.MaxUint32, WithChainUnaryInterceptor(firstInt, secondInt, lastInt))
+	defer func() {
+		cc.Close()
+		server.stop()
+	}()
+
+	var reply string
+	ctx := context.Background()
+	parentCtx := context.WithValue(ctx, ctxKey("parentKey"), 0)
+	if err := cc.Invoke(parentCtx, "/foo/bar", &expectedRequest, &reply); err != nil || reply != expectedResponse {
+		t.Fatalf("grpc.Invoke(_, _, _, _, _) = %v, want <nil>", err)
+	}
+}
+
+func (s) TestChainOnBaseUnaryClientInterceptor(t *testing.T) {
+	var (
+		parentKey  = ctxKey("parentKey")
+		baseIntKey = ctxKey("baseIntKey")
+	)
+
+	baseInt := func(ctx context.Context, method string, req, reply interface{}, cc *ClientConn, invoker UnaryInvoker, opts ...CallOption) error {
+		if ctx.Value(parentKey) == nil {
+			t.Fatalf("base interceptor should have %v in context", parentKey)
+		}
+		baseCtx := context.WithValue(ctx, baseIntKey, 1)
+		return invoker(baseCtx, method, req, reply, cc, opts...)
+	}
+
+	chainInt := func(ctx context.Context, method string, req, reply interface{}, cc *ClientConn, invoker UnaryInvoker, opts ...CallOption) error {
+		if ctx.Value(parentKey) == nil {
+			t.Fatalf("chain interceptor should have %v in context", parentKey)
+		}
+		if ctx.Value(baseIntKey) == nil {
+			t.Fatalf("chain interceptor should have %v in context", baseIntKey)
+		}
+		return invoker(ctx, method, req, reply, cc, opts...)
+	}
+
+	server, cc := setUpWithOptions(t, 0, math.MaxUint32, WithUnaryInterceptor(baseInt), WithChainUnaryInterceptor(chainInt))
+	defer func() {
+		cc.Close()
+		server.stop()
+	}()
+
+	var reply string
+	ctx := context.Background()
+	parentCtx := context.WithValue(ctx, ctxKey("parentKey"), 0)
+	if err := cc.Invoke(parentCtx, "/foo/bar", &expectedRequest, &reply); err != nil || reply != expectedResponse {
+		t.Fatalf("grpc.Invoke(_, _, _, _, _) = %v, want <nil>", err)
+	}
+}
+
+func (s) TestChainStreamClientInterceptor(t *testing.T) {
+	var (
+		parentKey    = ctxKey("parentKey")
+		firstIntKey  = ctxKey("firstIntKey")
+		secondIntKey = ctxKey("secondIntKey")
+	)
+
+	firstInt := func(ctx context.Context, desc *StreamDesc, cc *ClientConn, method string, streamer Streamer, opts ...CallOption) (ClientStream, error) {
+		if ctx.Value(parentKey) == nil {
+			t.Fatalf("first interceptor should have %v in context", parentKey)
+		}
```

Comment:
Let's also check `firstIntKey` to make sure it's not in context (interceptors are not accidentally called more than once).